ArtOfIllusion/src/artofillusion/procedural/Procedure.java

```diff
@@ -274,4 +278,74 @@ public void readFromStream(DataInputStream in, Scene theScene) throws IOExceptio
         to.getModule().setInput(to, from);
       }
   }
+  
+  public TextureParameter[] getTextureParameters(Object texture)
+  {
+    int count = 0;
+    for (Module mod : module)
+      if (mod instanceof ParameterModule)
+        count++;
+    
+    TextureParameter[] params = new TextureParameter[count];
+    count = 0;
+    for (Module mod : module)
+      if (mod instanceof ParameterModule)
+      {
+        params[count] = ((ParameterModule) mod).getParameter(texture);
+        ((ParameterModule) mod).setIndex(count++);
+      }
+    return params;
+  }
+
+  /** Determine whether given procedure texture has a non-zero value anywhere for a particular component.
+      @param component    the texture component to check for (one of the Texture *_COMPONENT constants)
+   */
+  public boolean hasTextureComponent(int component)
```

Comment:
This method now look not good for me as it implicitly depends of order and count of procedure output modules...

Reply:
It is specific to `Texture` procedures. As long as the procedure is a `Texture` procedure, it will work fine. That is why I think that this method and the `createTextureProcedure()` method do not belong in this class. I'm not sure where to put them though. For the time being, I would remove them entirely, and leave the original code in the Texture classes.